<!DOCTYPE html><html><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content="你好，我是米歐，歡迎來到我的部落格:)"><meta property="og:title" content="MMiooiMM"><meta property="og:description" content="你好，我是米歐，歡迎來到我的部落格:)"><meta property="og:image" content="https://raw.githubusercontent.com/MMiooiMM/MMiooiMM.github.io/master/images/VeryImportantFile.jpg"><title>C# 在迭代器中錯誤捕捉的問題 | MMiooiMM</title><link rel="stylesheet" type="text/css" href="/css/normalize.css"><link rel="stylesheet" type="text/css" href="/css/highlight.css"><link rel="stylesheet" type="text/css" href="/css/font.css"><link rel="stylesheet" type="text/css" href="/css/noise.css"><link rel="stylesheet" type="text/css" href="/css/mio.css"><link rel="stylesheet" type="text/css" href="//cdn.bootcss.com/font-awesome/4.5.0/css/font-awesome.min.css"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><meta name="generator" content="Hexo 4.2.1"></head><body><article class="wrapper"><div class="post-main"><div class="nav"><nav class="container"><a class="sidebar-nav-item active" href="/">Home</a><a class="sidebar-nav-item" href="/archives">Archives</a><a class="sidebar-nav-item" href="/about">About</a></nav><div class="container post-meta"><div class="post-tags"><a class="post-tag-link" href="/tags/CSharp/" rel="tag">CSharp</a></div><div class="post-time">2019-03-21</div></div></div><div class="container post-header"><h1>C# 在迭代器中錯誤捕捉的問題</h1></div><div class="container post-content"><p>練習 Linq 原始碼的時候，遇到了迭代器的錯誤處理問題，由迭代器產生的錯誤沒有被外層 <code>try-catch</code> 正確的捕捉，所以這篇會介紹問題並解答。</p>
<p>問題：在 <code>yield return</code> 或 <code>yield break</code> 前直接產生錯誤，如以下程式碼。</p>
<figure class="highlight csharp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">Main</span>(<span class="params"><span class="keyword">string</span>[] args</span>)</span> &#123;</span><br><span class="line">  <span class="keyword">try</span> &#123;</span><br><span class="line">    GetError();</span><br><span class="line">  &#125;</span><br><span class="line">  <span class="keyword">catch</span> (Exception ex) </span><br><span class="line">  &#123;  &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">static</span> IEnumerable&lt;<span class="keyword">int</span>&gt; <span class="title">GetError</span>(<span class="params"></span>)</span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">  <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">"獲得錯誤"</span>);</span><br><span class="line">  <span class="keyword">yield</span> <span class="keyword">break</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>這一個方法並不會讓程式直接停止運作，讓我想了很久，我的觀念以為錯誤會優先傳回，而這方法也因為沒有經過 <code>yield</code> 所以不會包裝成迭代器，所以我發了問題在<a href="https://ithelp.ithome.com.tw/questions/10192780" target="_blank" rel="noopener">C# 在迭代器中錯誤捕捉的問題 - iT 邦幫忙::一起幫忙解決難題，拯救 IT 人的一天</a>，不久就有好心的大大幫忙解決，詳細在連結裡，這邊就不多講了。</p>
<p>結論：使用關鍵字 <code>yield</code> 方法就會變成 <strong>lazy evaluation</strong>，就算沒有經過該段程式碼，並且在 <strong>lazy evaluation</strong> 中錯誤也需查詢才會發生。</p>
<hr>
<p>後記：我也嘗試使用反編譯工具(ILSpy)來看，看兩者之間到底差在哪裡，程式碼如下</p>
<figure class="highlight csharp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title">IteratorError</span></span><br><span class="line">&#123;</span><br><span class="line">  <span class="function"><span class="keyword">public</span> IEnumerable&lt;<span class="keyword">int</span>&gt; <span class="title">GetError</span>(<span class="params"></span>)</span></span><br><span class="line"><span class="function"></span>  &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">"獲得錯誤"</span>);</span><br><span class="line">    <span class="keyword">yield</span> <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">  &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">class</span> <span class="title">NormalError</span></span><br><span class="line">&#123;</span><br><span class="line">  <span class="function"><span class="keyword">public</span> IEnumerable&lt;<span class="keyword">int</span>&gt; <span class="title">GetError</span>(<span class="params"></span>)</span></span><br><span class="line"><span class="function"></span>  &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">"獲得錯誤"</span>);</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">null</span>;</span><br><span class="line">  &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>ILSpy 看到的原始碼為</p>
<figure class="highlight csharp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">internal</span> <span class="keyword">class</span> <span class="title">IteratorError</span></span><br><span class="line">&#123;</span><br><span class="line">  <span class="function"><span class="keyword">public</span> IEnumerable&lt;<span class="keyword">int</span>&gt; <span class="title">GetError</span>(<span class="params"></span>)</span></span><br><span class="line"><span class="function"></span>  &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">"獲得錯誤"</span>);</span><br><span class="line">    <span class="keyword">yield</span> <span class="keyword">break</span>;</span><br><span class="line">  &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">internal</span> <span class="keyword">class</span> <span class="title">NormalError</span></span><br><span class="line">&#123;</span><br><span class="line">  <span class="function"><span class="keyword">public</span> IEnumerable&lt;<span class="keyword">int</span>&gt; <span class="title">GetError</span>(<span class="params"></span>)</span></span><br><span class="line"><span class="function"></span>  &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">"獲得錯誤"</span>);</span><br><span class="line">  &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><code>IteratorError</code> 的 <code>GetError()</code> 的 <code>yield return 0</code> 變成了 <code>yield break</code>，而 <code>NormalError</code> 的 <code>GetError()</code> 的 <code>return</code> 直接消失。</p>
<p>我認為是 IDE 產生的幻覺，因為程式碼跑不到的地方會有提示告訴開發者，但編譯後的樣子卻能發現兩者的差異，所以不經過 <code>yield</code> 也會變成 <strong>lazy evaluation</strong>，只是你以為編譯後會一樣。</p>
<hr>
<p>再更，用 Ildasm.exe (IL 反組譯工具)來查看中繼碼，而不是像上面的還原成 C# 碼。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">.method public hidebysig instance class [mscorlib]System.Collections.Generic.IEnumerable&#96;1&lt;int32&gt; </span><br><span class="line">        GetError() cil managed</span><br><span class="line">&#123;</span><br><span class="line">  .custom instance void [mscorlib]System.Runtime.CompilerServices.IteratorStateMachineAttribute::.ctor(class [mscorlib]System.Type) &#x3D; ( 01 00 28 43 6F 6E 73 6F 6C 65 41 70 70 32 2E 49   &#x2F;&#x2F; ..(ConsoleApp2.I</span><br><span class="line">                                                                                                                                        74 65 72 61 74 6F 72 45 72 72 6F 72 2B 3C 47 65   &#x2F;&#x2F; teratorError+&lt;Ge</span><br><span class="line">                                                                                                                                        74 45 72 72 6F 72 3E 64 5F 5F 30 00 00 )          &#x2F;&#x2F; tError&gt;d__0..</span><br><span class="line">  &#x2F;&#x2F; 程式碼大小       15 (0xf)</span><br><span class="line">  .maxstack  8</span><br><span class="line">  IL_0000:  ldc.i4.s   -2</span><br><span class="line">  IL_0002:  newobj     instance void ConsoleApp2.IteratorError&#x2F;&#39;&lt;GetError&gt;d__0&#39;::.ctor(int32)</span><br><span class="line">  IL_0007:  dup</span><br><span class="line">  IL_0008:  ldarg.0</span><br><span class="line">  IL_0009:  stfld      class ConsoleApp2.IteratorError ConsoleApp2.IteratorError&#x2F;&#39;&lt;GetError&gt;d__0&#39;::&#39;&lt;&gt;4__this&#39;</span><br><span class="line">  IL_000e:  ret</span><br><span class="line">&#125; &#x2F;&#x2F; end of method IteratorError::GetError</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">.method public hidebysig instance class [mscorlib]System.Collections.Generic.IEnumerable&#96;1&lt;int32&gt; </span><br><span class="line">        GetError() cil managed</span><br><span class="line">&#123;</span><br><span class="line">  &#x2F;&#x2F; 程式碼大小       12 (0xc)</span><br><span class="line">  .maxstack  1</span><br><span class="line">  .locals init ([0] class [mscorlib]System.Collections.Generic.IEnumerable&#96;1&lt;int32&gt; V_0)</span><br><span class="line">  IL_0000:  nop</span><br><span class="line">  IL_0001:  ldstr      bytearray (72 73 97 5F 2F 93 A4 8A )                         &#x2F;&#x2F; rs._&#x2F;...</span><br><span class="line">  IL_0006:  newobj     instance void [mscorlib]System.Exception::.ctor(string)</span><br><span class="line">  IL_000b:  throw</span><br><span class="line">&#125; &#x2F;&#x2F; end of method NormalError::GetError</span><br></pre></td></tr></table></figure>

<p>兩段程式碼長得更不一樣了，當然不只 GetError 方法不一樣，迭代器版本的類別也多了一個屬於 GetError 的迭代器類別，可以很清楚區分兩者差異。</p>
<p>參考資料<br><br><a href="http://kevintsengtw.blogspot.com/2011/09/netilspy-telerik-justdecompile.html" target="_blank" rel="noopener">mrkt 的程式學習筆記: .NET反組譯工具：ILSpy, Telerik JustDecompile</a><br><br><a href="https://docs.microsoft.com/zh-tw/dotnet/framework/tools/ildasm-exe-il-disassembler" target="_blank" rel="noopener">Ildasm.exe (IL 反組譯工具) Microsoft Docs</a></p>
</div></div><div class="post-main post-comment"><div id="disqus_thread"></div><script type="text/javascript">
var disqus_shortname = 'mio-blog';
var disqus_identifier = '2019/03/21/2019-03-21-csharp-iterator-error-catch-question/';
var disqus_title = 'C# 在迭代器中錯誤捕捉的問題';
var disqus_url = 'http://yoursite.com/2019/03/21/2019-03-21-csharp-iterator-error-catch-question/';
(function() {
   var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
   dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
   (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript" target="_blank" rel="noopener">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" target="_blank" rel="noopener" class="dsq-brlink">Blog comments powered by <span class="logo-disqus">Disqus</span></a></div></article><link rel="stylesheet" type="text/css" href="//cdn.bootcss.com/fancybox/2.1.5/jquery.fancybox.css"><script src="//cdn.bootcss.com/jquery/2.0.3/jquery.min.js"></script><script src="//cdn.bootcss.com/fancybox/2.1.5/jquery.fancybox.pack.js"></script><script>$(document).ready(function() {
  $(".fancybox").fancybox();
});
</script><script>(function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
e=o.createElement(i);r=o.getElementsByTagName(i)[0];
e.src='//www.google-analytics.com/analytics.js';
r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
ga('create','UA-156431110-1');ga('send','pageview');</script></body></html>